strawberry/permission.py

```diff
@@ -1,21 +1,162 @@
+
+    error_class: Type[GraphQLError] = StrawberryGraphQLError
+
+    _schema_directive: Optional[StrawberrySchemaDirective] = None
+
+    @abc.abstractmethod
     def has_permission(
         self, source: Any, info: Info, **kwargs
     ) -> Union[bool, Awaitable[bool]]:
         raise NotImplementedError(
             "Permission classes should override has_permission method"
         )
+
+    def handle_no_permission(self) -> None:
+        """
+        Default error raising for permissions.
+        This can be overridden to customize the behavior.
+        """
+
+        # Instantiate error class
+        error = self.error_class(self.message or "")
+
+        if self.error_extensions:
+            # Add our extensions to the error
+            if not error.extensions:
+                error.extensions = dict()
+            error.extensions.update(self.error_extensions)
+
+        raise error
+
+    @property
+    def schema_directive(self) -> Optional[StrawberrySchemaDirective]:
+        if not self._schema_directive:
+            self._schema_directive = StrawberrySchemaDirective(
+                self.__class__.__name__,
+                self.__class__.__name__,
+                [Location.FIELD_DEFINITION],
+                [],
+            )
+        return self._schema_directive
+
+
+class PermissionExtension(FieldExtension):
+    """
+    Handles permissions for a field
+    Instantiate this as a field extension with all of the permissions you want to apply
+
+    fail_silently: bool = False will return None or [] if the permission fails
+    instead of raising an exception. This is only valid for optional or list fields.
+
+    NOTE:
+    Currently, this is automatically added to the field, when using
+    field.permission_classes
+    This is deprecated behavior, please manually add the extension to field.extensions
+    """
+
+    def __init__(self, permissions: List[BasePermission], fail_silently: bool = False):
+        self.permissions = permissions
+        self.fail_silently = fail_silently
+        self.return_empty_list = False
+
+    def apply(self, field: StrawberryField) -> None:
+        """Applies all of the permission directives to the schema"""
+        for permission in self.permissions:
+            if permission.schema_directive:
+                cast(List, field.directives).append(permission.schema_directive)
+        # We can only fail silently if the field is optional or a list
+        if self.fail_silently:
+            if isinstance(field.type, StrawberryOptional):
+                if isinstance(field.type.of_type, StrawberryList):
+                    self.return_empty_list = True
+            elif isinstance(field.type, StrawberryList):
+                self.return_empty_list = True
+            else:
+                raise Exception(
+                    "Cannot use fail_silently=True with a non-optional "
+                    "or non-list field"
+                )
```

Comment:
Should I move this to a custom exception in the exceptions module, or is it okay like this (e.g. with a value error)?
How can I get the fancy diagnostics for this?

Reply:
I think you just need to create a subclass of `StrawberryException` and raise it. You can take a look at some examples like `InnvalidArgumentTypeError`